File: CPP/ByTier/O/1662_O_Circular_Maze.cpp
// Problem: CF 1662 O - Circular Maze
// https://codeforces.com/contest/1662/problem/O

/*
O. Circular Maze
Time Complexity: O(n * 360) per test case, where n is the number of walls.
Space Complexity: O(40 * 360) = O(1) due to fixed dimensions.

Algorithms/Techniques:
- This problem uses a graph traversal approach (DFS) on a discretized circular maze.
- The maze is modeled as a grid where each cell represents a sector in the circular space.
- Walls are marked as blocked cells.
- Then, a DFS from the center (1,1) checks if there's a path to the outer boundary (y=720).

The maze is converted into a 2D grid by mapping radius and angle to indices.
Each circular wall is discretized into sectors. Straight walls are also converted
into sectors. Then, DFS is run from (1,1) to check if we can reach the border.
*/
#include <stdio.h>
#include <string.h>
#include <iostream>

#define use 1, n, rt
#define left l, mid, ls[p]
#define righ mid + 1, r, rs[p]
#define N 500000
#define fer(x, y, z) for (register int x = y; x <= z; x++)
using namespace std;
int t, n;
bool pd[41][361 * 2], vis[41][361 * 2]; // pd marks walls, vis marks visited
int sak_dfs(int x, int y) {
  if (x == 40) { // If reached outermost layer
    return true;
  }
  vis[x][y] = 1; // Mark current position as visited
  int flag = 0;
  if (y == 1) { // Wrap around to end
    if (!pd[x][720] && !vis[x][720])
      flag += sak_dfs(x, 720);
  } else if (!pd[x][y - 1] && !vis[x][y - 1]) // Move left (decrease angle)
    flag += sak_dfs(x, y - 1);
  if (y == 720) { // Wrap around to start
    if (!pd[x][1] && !vis[x][1])
      flag += sak_dfs(x, 1);
  } else if (!pd[x][y + 1] && !vis[x][y + 1]) // Move right (increase angle)
    flag += sak_dfs(x, y + 1);
  if (x != 1 && !pd[x - 1][y] && !vis[x - 1][y]) // Move up (decrease radius)
    flag += sak_dfs(x - 1, y);
  if (!pd[x + 1][y] && !vis[x + 1][y]) // Move down (increase radius)
    flag += sak_dfs(x + 1, y);
  return flag != 0;
}
signed main() {
  cin >> t;
  while (t--) {
    char ch;
    int x, y, z;
    cin >> n;
    memset(pd, 0, sizeof pd), memset(vis, 0, sizeof vis); // Reset for each test case
    fer(i, 1, n) {
      scanf(" %c%d%d%d", &ch, &x, &y, &z);
      x *= 2; // Scale up to improve precision
      y *= 2, z *= 2;
      if (ch == 'C') { // Circular wall
        y += 2, z += 2; // Adjust for scaling and shift angles to align with indexing
        if (y > z) { // Overlapping case
          fer(i, y, 720) pd[x][i] = 1; // Mark the portion from y to end
          fer(i, 1, z) pd[x][i] = 1;   // Then mark from start to z
        } else
          fer(i, y, z) pd[x][i] = 1; // Mark regular segment
      } else // Straight wall
        fer(i, x, y) pd[i][z + 2] = 1; // Mark vertical wall at angle z+2 (scaled)
    }
    if (sak_dfs(1, 1)) // Start DFS from center cell (1,1)
      cout << "YES\n";
    else
      cout << "NO\n";
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/O/1662_O_Circular_Maze.py
# Problem: CF 1662 O - Circular Maze
# https://codeforces.com/contest/1662/problem/O

"""
O. Circular Maze

Algorithms/Techniques:
- Graph traversal (BFS)
- Grid representation of circular maze with angle and radius coordinates
- Handling wraparound angles using modulo arithmetic
- Using sets for efficient neighbor lookups

Time Complexity: O(MAX_RADIUS * 360) = O(1) since MAX_RADIUS is fixed at 20
Space Complexity: O(MAX_RADIUS * 360) = O(1) due to fixed dimensions

The problem models a circular maze where we need to find a path from center (radius 0) 
to the outer boundary (radius MAX_RADIUS - 1), avoiding walls. Walls are either:
- Circular: defined by radius and two angles (clockwise arc)
- Straight: defined by angle and radial range

We use BFS traversal over a grid where each cell represents (radius, angle). 
Each cell has up to 4 neighbors (up, right, left, down), with wrapping around 
angles and boundaries checked against wall positions.
"""

from itertools import chain, islice
from sys import stdin

MAX_RADIUS = 20
lines = iter(stdin.readlines()[1:])
for line in lines:
    n = int(line)
    # 2D arrays to mark walls: circular_wall[r][theta] for circular walls,
    # straight_wall[r][theta] for straight walls
    circular_wall, straight_wall = (
        [[False] * 360 for _ in range(MAX_RADIUS)] for _ in range(2)
    )
    for shape, *params in map(str.split, islice(lines, n)):
        params = map(int, params)
        if shape == "C":
            r, theta_1, theta_2 = params
            r -= 1  # Convert to 0-based indexing
            # Handle wraparound angle ranges correctly
            theta_range = (
                range(theta_1, theta_2)
                if theta_1 <= theta_2
                else chain(range(theta_1, 360), range(0, theta_2))
            )
            for theta in theta_range:
                circular_wall[r][theta] = True
        else:
            assert shape == "S"
            r1, r2, theta = params
            r1 -= 1  # Convert to 0-based indexing
            r2 -= 1
            # Mark all radii from r1 to r2-1 at angle theta as walls
            for r in range(r1, r2):
                straight_wall[r][theta] = True
    # Start BFS from all valid positions at radius 0 (center)
    queue = [(0, i) for i, inner_wall in enumerate(circular_wall[0]) if not inner_wall]
    seen = set(queue)
    while queue:
        row, col = queue.pop()
        neighbors = []
        # Check up neighbor (if valid and not a circular wall)
        if row >= 1 and not circular_wall[row][col]:
            neighbors.append((row - 1, col))
        # Right neighbor with wraparound
        right_col = (col + 1) % 360
        if not straight_wall[row][right_col]:
            neighbors.append((row, right_col))
        # Left neighbor with wraparound
        if not straight_wall[row][col]:
            neighbors.append((row, (col - 1) % 360))
        # Down neighbor (if valid and not a circular wall)
        next_row = row + 1
        if not circular_wall[next_row][col]:
            if next_row == MAX_RADIUS - 1:
                print("YES")
                break
            neighbors.append((next_row, col))
        for neighbor in neighbors:
            if neighbor in seen:
                continue
            queue.append(neighbor)
            seen.add(neighbor)
    else:
        print("NO")


# https://github.com/VaHiX/codeForces/